# Regressão Logística

Universidade Federal de Campina Grande<br/>
Programa de Pós-Graduação em Ciência da Computação<br/>
Disciplina: Aprendizagem de Máquina<br/>
Aluno: Ruan Victor Bertoldo Reis de Amorim

In [1]:
from sklearn import linear_model
import numpy as np
import time
import pandas as pd

## Função logística (Sigmóide)

In [2]:
def logistic(w, X):
    return 1 / (1 + np.exp(-np.dot(X, w)))

## Gradiente ascendente (Vetorizado)

In [3]:
def step_gradient_vectorized(w_current, X, Y, learningRate):
    gradient = np.zeros((len(w_current), 1))
    
    for var in range(len(gradient)):
        gradient[var] += np.dot(X[:,var], Y - logistic(w_current, X))
    
    w_current += (gradient * learningRate)
    return [w_current, gradient]

## Iterações

In [4]:
def gradient_ascendant_runner_vectorized(starting_w, X, Y, learning_rate, epsilon):
    w = starting_w
    gradient = np.array([np.inf, np.inf, np.inf, np.inf, np.inf])

    while (np.linalg.norm(gradient) >= epsilon):
        w, gradient = step_gradient_vectorized(w, X, Y, learning_rate)
    return w

## Função de predição com limiar de 0.5

In [5]:
def predict(w, X):
    logit = logistic(w, X)
    return np.greater_equal(logit, 0.5).astype(int)

## Ponto de partida - Leitura dos dados, configurações e resultados

In [6]:
points = pd.read_csv("dados/iris.csv")
points['iris'] = np.where(points['iris'] == 'Iris-setosa', 0, 1)

X = np.c_[np.ones(len(points)), points[['sepal-length', 'sepal-width', 'petal-length', 'petal-width']]]
Y = points[['iris']]

init_w = np.zeros((5,1))

learning_rate = 0.0001
epsilon = 3.0

print("Iniciando gradiente ascendente com w0 = {0}, w1 = {1}, w2 = {2}, w3 = {3}, w4 = {4}"
      .format(init_w[0], init_w[1], init_w[2], init_w[3], init_w[4]))
print("Executando...")

tic = time.time()
w = gradient_ascendant_runner_vectorized(init_w, X, Y, learning_rate, epsilon)
toc = time.time()

print("\nGradiente ascendente convergiu com w0 = {0}, w1 = {1}, w2 = {2}, w3 = {3}, w4 = {4}".
      format(w[0], w[1], w[2], w[3], w[4]))
print("\nTempo de Execução: " + str((toc-tic)) + " segundos")


Iniciando gradiente ascendente com w0 = [ 0.], w1 = [ 0.], w2 = [ 0.], w3 = [ 0.], w4 = [ 0.]
Executando...

Gradiente ascendente convergiu com w0 = [-0.27069865], w1 = [-0.42693718], w2 = [-1.43042757], w3 = [ 2.25546687], w4 = [ 1.02373226]

Tempo de Execução: 5.460176944732666 segundos


## Predições com os dados de treino

In [17]:
predictions = np.c_[predict(w, X), Y]
result = ""

for pred in predictions:
    result += str(pred) + ","
    
print("PREDIÇÕES:")
print(result)

PREDIÇÕES:
[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[0 0],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],[1 1],


## Coeficientes da regressão logística com Scikit Learn

In [245]:
logreg = linear_model.LogisticRegression()
model = logreg.fit(X, points.iris)
print (model.coef_)

[[-0.25478973 -0.38180457 -1.43873414  2.26192475  1.02592844]]


### Os coeficientes encontrados pelo nosso algoritmo são próximos aos coeficientes do Scikit Learn:

**Scikit Learn**<br/>
[-0.25478973, -0.38180457, -1.43873414, 2.26192475, 1.02592844] <br/><br/>
**Nosso Algoritmo**<br/>
[-0.27069865, -0.42693718, -1.43042757, 2.25546687, 1.02373226]